lib/Sema/CSSimplify.cpp

```diff
@@ -4307,6 +4307,14 @@ ConstraintSystem::matchTypesBindTypeVar(
     }
   }
 
+  if (type->is<PackType>() || type->is<PackArchetypeType>()) {
+    if (!typeVar->getImpl().canBindToPack())
+      return getTypeMatchFailure(locator);
```

Comment:
I think we need a bi-directional fix here similar to `fixReferenceMismatch` otherwise constraint solver is just going to fail with "failed to produce a diagnostic" because it wouldn't be able to form a valid solution due to this binding always failing.

Reply:
I'll let @hborla handle this one :) I was just rebasing this PR so that she could finish it off.

Reply:
Yes, I'm going to pick this up and address the feedback about diagnostics!